File: BACKEND/SCRIPT/BIOASSAY_ONTO/wh_bioassay_onto.py
import os
import shutil
import subprocess
import re
from datetime import datetime
import sys
sys.path.append(os.getenv('TG_DIR')+"/BACKEND/SCRIPT/LIB_PYTHON/")
import time
from datetime import datetime
import smtplib
import psycopg2
import pickle
import pprint
from fct_utils import *
from loader import *


job_name = 'wh_bioassay_onto'
JOB_ID = get_job_id_by_name(job_name,True)
JOB_INFO=GLB_TREE[JOB_ID]
PROCESS_CONTROL['JOB_NAME'] =job_name

TG_DIR = os.getenv('TG_DIR')

if TG_DIR is False:
    print('NO TG_DIR found')
    exit()

if not os.path.isdir(TG_DIR):
    print('TG_DIR value is not a directory', TG_DIR)
    exit()


add_log("Check directory")

# Get parent job info
CK_SEQ_ONTOL_INFO = GLB_TREE[get_job_id_by_name('ck_bioassay_onto')]


# Get to the directory set by ck_bioassay_onto
W_DIR = os.path.join(TG_DIR, GLB_VAR['PROCESS_DIR'], JOB_INFO['DIR'], CK_SEQ_ONTOL_INFO['TIME']['DEV_DIR'])
if not os.path.isdir(W_DIR):
    fail_process(JOB_ID + "002", 'NO ' + W_DIR + ' found')
os.chdir(W_DIR)
if os.getcwd() != W_DIR:
    fail_process(JOB_ID + "003", 'Unable to chdir to ' + W_DIR)

# Assign the directory to the process control, so the next job knows where to look
PROCESS_CONTROL['DIR'] = CK_SEQ_ONTOL_INFO['TIME']['DEV_DIR']

# Check if the file bao_complete_merged.owl is present
F_FILE = os.path.join(W_DIR, 'bao_complete_merged.owl')
if not check_file_exist(F_FILE):
    fail_process(JOB_ID + "004", 'NO ' + F_FILE + ' found')

STATS = {'ENTRY': 0}


add_log("Load data from database")


# We are going to load the data from the database
DATA = {}
# MAX_DBID is the maximum bioassay_onto_entry_id in the database
MAX_DBID = 0
res = run_query("SELECT bioassay_onto_entry_id,bioassay_tag_id,bioassay_label,bioassay_definition FROM bioassay_onto_entry ORDER BY bioassay_tag_id ASC")
if res is False:
    fail_process(JOB_ID + "007", 'Unable to fetch from database')

for tab in res:
    DATA[tab['bioassay_tag_id']] = {
        'DB': tab['bioassay_onto_entry_id'],
        'NAME': tab['bioassay_label'],
        'DESC': tab['bioassay_definition'],
        'STATUS': 'FROM_DB',# STATUS is used to determine if the entry is new, updated or from the database
        'EXTDB': [],
        'CHILD': {}
    }
    # Update the MAX_DBID if the current bioassay_onto_entry_id is greater than the current MAX_DBID
    if tab['bioassay_onto_entry_id'] > MAX_DBID:
        MAX_DBID = tab['bioassay_onto_entry_id']

add_log("Load data from file")

# Load data from the owl file
ROOTS = {}
# This is an owl file, which is an xml file
# We are going to read the file line by line
with open('bao_complete_merged.owl', 'r') as fp:
    while True:
        line = fp.readline()
        if not line:
            break
        # Each record start with Class
        if '<owl:Class' not in line:
            continue
        
        # Extract the ID from the line
        P1 = line.rfind('/')
        P2 = line.rfind('"')
        ID = line[P1+5:P2]
        
        if P1 is False or P2 is False:
            continue
        # We are only interested in the BAO_ entries
        if not ID.startswith('BAO_'):
            continue
        
        # Initialize the variables
        STATS['ENTRY'] += 1
        N_OPEN = 0
        DESC = []
        NAME = ''
        PARENTS = []
        VALID = True
        EXTDB = []

		# Since a record is a set of lines, we are going to read the lines until we reach the end of the record
        while True:
            line = fp.readline()
            if not line:
                break
            if 'IAO_0000115' in line:
                #Description text
                DESC.append(get_in_between(line, 'obo:IAO_0000115') + "\n")
            if '<rdfs:label' in line:
                NAME = get_in_between(line, 'rdfs:label')
                
			# We don't want deprecated entries
            if 'owl:deprecated' in line:
                VALID = False
                break
            
            # Get the parent
            if '<rdfs:subClassOf' in line:
                tstr = get_tag(line, 'rdf:resource')
                PARENT = tstr[tstr.rfind('/')+1:]
                if PARENT.startswith('bao#BAO'):
                    PARENT = PARENT[4:]
                    if PARENT != '' and PARENT != 'owl#Thing':
                        PARENTS.append(PARENT)
            if '<owl:Class' in line:
                N_OPEN += 1
            if '</owl:Class' in line:
                if N_OPEN == 0:
                    break
                else:
                    N_OPEN -= 1
                    
		# If the record is Deprecated -> continue
        if not VALID:
            continue
		
        # Not existing in the database -> create the record
        if ID not in DATA:
            
            MAX_DBID += 1
            DATA[ID] = {
                'DB': MAX_DBID,
                'NAME': NAME,
                'DESC': DESC,
                'STATUS': 'TO_INS',
                'EXTDB': EXTDB,
                'CHILD': {}
            }
        # Some records in DATA can be created because of the child/parent, but don't have a DB ID, in that case, we create the record
        elif ID in DATA and 'DB' not in DATA[ID]:
            MAX_DBID += 1
            DATA[ID]['DB'] = MAX_DBID
            DATA[ID]['NAME'] = NAME
            DATA[ID]['DESC'] = DESC
            DATA[ID]['STATUS'] = 'TO_INS'
            DATA[ID]['EXTDB'] = EXTDB
        else:
            # The record exist -> set to valid, unless some data has changed
            DATA[ID]['STATUS'] = 'VALID'
            if NAME != DATA[ID]['NAME']:
                DATA[ID]['NAME'] = NAME
                DATA[ID]['STATUS'] = 'TO_UPD'
            if ''.join(DESC) != DATA[ID]['DESC']:
                DATA[ID]['DESC'] = DESC
                DATA[ID]['STATUS'] = 'TO_UPD'

		# That record has no parent -> Root 
		# We are going to use this information to create the nested set representation
        if not PARENTS:
            ROOTS[ID] = True
        else:
            for C in PARENTS:
                # This is causing the creation of a record in $DATA without name/description. See case above
                DATA[C]['CHILD'][ID] = True

add_log("Create nested set representation")

if not ROOTS:
    fail_process(JOB_ID + "009", 'No root found')

with open('TREE.csv', 'w') as fp:
    VALUE = 0


	# Create nested set representation that is going to assign boundary numbers.
	# Let's say that the root has for boundary 1 10.
	# The two childs:  A 2-5 and B 6-9
	# And the A has a child C 3-4
	# If we want ALL parents of C, we are going to look outside the boundaries, i.e. <3 for the left side and >4 for the right side.
	# By doing so we get A 2-5 and root 1-10 but not B because the left boundary 6 is above C left boundary.
	# Similarly, if we want children of Root, we will look inside the boundaries i.e >1 for theleft side and <10 for the right side, leading to A B and C.
    def gen_tree(DATA, ROOTS, LEVEL,LEVEL_V):
        LEVEL += 1
        for RID in ROOTS.keys():
            if RID not in DATA:
                print(RID)
                continue
            LEVEL_V += 1
            LEVEL_LEFT = LEVEL_V
            if 'CHILD' in DATA[RID]:
                print(RID+" "+str(DATA[RID]['CHILD'])+" "+str(LEVEL)+" "+str(LEVEL_V))
                LEVEL_V=gen_tree(DATA, DATA[RID]['CHILD'], LEVEL, LEVEL_V)
            LEVEL_V += 1
            LEVEL_RIGHT = LEVEL_V
            fp.write(f"{DATA[RID]['DB']}\t{LEVEL}\t{LEVEL_LEFT}\t{LEVEL_RIGHT}\n")
        return LEVEL_V
    gen_tree(DATA, ROOTS, 0, VALUE)

	# $DATA now contains both the data from the database and from the file
	# So we look at each of those to see what needs to be deleted, updated, inserted

print("Pushing changes to database")


#print(DATA['BAO_0002235'])
#exit()

# Open the file that is going to contain the new records
with open('ENTRY.csv', 'w') as fpE:
    HAS_NEW_DATA = False
    #We are going to look at each record and see what we need to do
    for ID, INFO in DATA.items():
        # FROM_DB => record is not in the file anymore -> DELETE IT	
        if INFO['STATUS'] == 'FROM_DB':
            print(ID)
            print(INFO)
            QUERY = f"DELETE FROM bioassay_onto_entry WHERE bioassay_onto_entry_id = {INFO['DB']}"
            if not run_query_no_res(QUERY):
                fail_process(JOB_ID + "012", 'Unable to run query ' + QUERY)
        # Update the record
        elif INFO['STATUS'] == 'TO_UPD':
            print(INFO)
            QUERY = f"UPDATE bioassay_onto_entry SET bioassay_tag_id = '{prep_string(ID)}', bioassay_label = '{prep_string(INFO['NAME'])}', bioassay_definition = '{prep_string(''.join(INFO['DESC']))}' WHERE bioassay_onto_entry_id = {INFO['DB']}"
            print(QUERY)
            if not run_query_no_res(QUERY):
                fail_process(JOB_ID + "013", 'Unable to run query ' + QUERY)
        elif INFO['STATUS'] == 'TO_INS':
            HAS_NEW_DATA = True
            fpE.write(f"{INFO['DB']}\t{ID}\t\"{INFO['NAME']}\"\t\"{''.join(INFO['DESC'])}\"\n")

print("Pushing new records to database")

if HAS_NEW_DATA:
    command = f"\COPY {GLB_VAR['DB_SCHEMA']}.bioassay_onto_entry(bioassay_onto_entry_id,bioassay_tag_id,bioassay_label,bioassay_definition) FROM 'ENTRY.csv' (DELIMITER E'\\t', null \\\"NULL\\\" ,format CSV )"
    print(f"{DB_INFO['COMMAND']} -c \"{command}\"")
    return_code=os.system(f"{DB_INFO['COMMAND']} -c \"{command}\"")
    if return_code != 0:
        fail_process(JOB_ID + "014", 'Unable to insert bioassay entry')

print("Delete content of bioassay_onto_hierarchy")

if not run_query_no_res("TRUNCATE TABLE bioassay_onto_hierarchy"):
    fail_process(JOB_ID + "015", 'Unable to truncate bioassay_onto_hierarchy')

print("Pushing tree to database")

command = f"\COPY {GLB_VAR['DB_SCHEMA']}.bioassay_onto_hierarchy(bioassay_onto_entry_id,bioassay_onto_level,bioassay_onto_level_left,bioassay_onto_level_right) FROM 'TREE.csv' (DELIMITER E'\\t', null \\\"NULL\\\" ,format CSV, HEADER )"
print(f"{DB_INFO['COMMAND']} -c \"{command}\"")
return_code=os.system(f"{DB_INFO['COMMAND']} -c \"{command}\"")
if return_code != 0:
    fail_process(JOB_ID + "016", 'Unable to insert tree')

update_stat('bioassay_onto_entry', 'bioassay_onto', STATS['ENTRY'], JOB_ID)

print("Delete obsolete files")
list_files = ['TREE.csv', 'ENTRY.csv']
for F in list_files:
    if os.path.isfile(F):
        os.unlink(F)

print("Push to prod")
push_to_prod(JOB_INFO,W_DIR)


success_process()

File: BACKEND/SCRIPT/BIOASSAY_ONTO/wh_bioassay_onto.php
<?php
$JOB_NAME='wh_bioassay_onto';
$TG_DIR= getenv('TG_DIR');
if ($TG_DIR===false)  die('NO TG_DIR found ');
if (!is_dir($TG_DIR)) die('TG_DIR value is not a directory '.$TG_DIR);
require_once($TG_DIR.'/BACKEND/SCRIPT/LIB/loader.php');


$JOB_ID=getJobIDByName($JOB_NAME);



addLog("Check directory");

	
	/// Get parent job info
	$CK_SEQ_ONTOL_INFO=$GLB_TREE[getJobIDByName('ck_bioassay_onto')];

	/// Get job info
	$JOB_INFO=$GLB_TREE[$JOB_ID];

	/// Get to the directory set by ck_bioassay_onto
	$W_DIR=$TG_DIR.'/'.$GLB_VAR['PROCESS_DIR'].'/'.$JOB_INFO['DIR'].'/'.$CK_SEQ_ONTOL_INFO['TIME']['DEV_DIR'];	
	if (!is_dir($W_DIR)) 															failProcess($JOB_ID."002",'NO '.$W_DIR.' found ');
	if (!chdir($W_DIR))																failProcess($JOB_ID."003",'Unable to chdir to '.$W_DIR);
	
	/// We assign the directory to the process control, so the next job knows where to look
	$PROCESS_CONTROL['DIR']=$CK_SEQ_ONTOL_INFO['TIME']['DEV_DIR'];

	/// We need to check if the file bao_complete_merged.owl is present
	$F_FILE=$W_DIR.'/bao_complete_merged.owl';	if (!checkFileExist($F_FILE))		failProcess($JOB_ID."004",'NO '.$F_FILE.' found ');
	
	

	$STATS=array('ENTRY'=>0);

	
addLog("Load data from database");
	/// We are going to load the data from the database
	$DATA=array();
	/// MAX_DBID is going to be used to create new records
	$MAX_DBID=0;
	$res=runQuery("SELECT bioassay_onto_entry_id,bioassay_tag_id,bioassay_label,bioassay_definition 
	 FROM bioassay_onto_entry");
	if ($res===false)																failProcess($JOB_ID."007",'Unable to fetch from database');
	
	foreach ($res as $tab)
	{
		$DATA[$tab['bioassay_tag_id']]=array(
			'DB'=>$tab['bioassay_onto_entry_id'],
			'NAME'=>$tab['bioassay_label'],
			'DESC'=>$tab['bioassay_definition'],
			'STATUS'=>'FROM_DB', /// STATUS is used to know if the record is in the file or not
		'EXTDB'=>array());
		/// Update MAX_DBID so we can create new records
		 if ($tab['bioassay_onto_entry_id']>$MAX_DBID)$MAX_DBID=$tab['bioassay_onto_entry_id'];
	}

addLog("Load data from file");
	/// This is an owl file, which is an xml file
	/// However, we are going to treat it as a text file
	$fp=fopen('bao_complete_merged.owl','r');if (!$fp)										failProcess($JOB_ID."008",'Unable to open bao_complete_merged.owl');
	$ROOTS=array();
	while(!feof($fp))
	{
		$line=stream_get_line($fp,1000,"\n");
		/// Each record start with class
		if (strpos($line,'<owl:Class')===false)continue;
		
		/// Extracting the ID:
		$P1=strrpos($line,'/');
		$P2=strrpos($line,'"');
		$ID=substr($line,$P1+5,$P2-$P1-5);
		if ($P1===false||$P2===false)continue;
		/// We only want BAO records
		if (substr($ID,0,4)!='BAO_'){continue;}
			
		$STATS['ENTRY']++;
		
		/// Initialize variables
		$N_OPEN=0;
		$DESC=array();
		$NAME='';
		$PARENTS=array();
		$VALID=true;
		$EXTDB=array();
		/// Since a record is a set of lines, we read all lines until we get to </owl:Class
		do
		{
			$line=stream_get_line($fp,1000,"\n");
			///Description text
			if (strpos($line,'IAO_0000115')!==false)$DESC[]=getInBetween($line,'obo:IAO_0000115')."\n";
			if (strpos($line,'<rdfs:label')!==false)$NAME=getInBetween($line,'rdfs:label');
			
			/// We don't want deprecated entries
			if (strpos($line,'owl:deprecated')!==false){$VALID=false;break;}

			/// Get the parent
			if (strpos($line,'<rdfs:subClassOf')!==false)
			{
				$tstr=getTag($line,'rdf:resource');
				$PARENT=substr($tstr,strrpos($tstr,'/')+1);
				if (substr($PARENT,0,7)!='bao#BAO')continue;
				$PARENT=substr($PARENT,4);
				if ($PARENT!=''&& $PARENT!='owl#Thing' )$PARENTS[]=$PARENT;
			
			}
			
			
			if (strpos($line,'<owl:Class')!==false)++$N_OPEN;
			if (strpos($line,'</owl:Class')!==false)
			{
				if ($N_OPEN==0)				break;
				else --$N_OPEN;
			}
		}while(!feof($fp));

		//Deprecated -> continue
		if (!$VALID)continue;

		// Not existing in the database -> create the record
		if (!isset($DATA[$ID]))
		{
			++$MAX_DBID;
			$DATA[$ID]=array('DB'=>$MAX_DBID,'NAME'=>$NAME,'DESC'=>$DESC,'STATUS'=>'TO_INS','EXTDB'=>$EXTDB);
			
		}
		/// Some records in DATA can be created because of the child/parent, but don't have a DB ID, in that case, we create the record
		else if (isset($DATA[$ID]) && !isset($DATA[$ID]['DB']))
		{
			++$MAX_DBID;
			$DATA[$ID]['DB']=$MAX_DBID;
			$DATA[$ID]['NAME']=$NAME;
			$DATA[$ID]['DESC']=$DESC;
			$DATA[$ID]['STATUS']='TO_INS';
			$DATA[$ID]['EXTDB']=$EXTDB;
		}
		else 
		{
			/// The record exist -> set to valid, unless some data has changed
			$DATA[$ID]['STATUS']='VALID';
			if ($NAME!=$DATA[$ID]['NAME']){
				$DATA[$ID]['NAME']=$NAME;
				
				$DATA[$ID]['STATUS']='TO_UPD';}
			if (implode("",$DESC)!=$DATA[$ID]['DESC']){
				
				$DATA[$ID]['DESC']=$DESC;
				

				$DATA[$ID]['STATUS']='TO_UPD';}
		
		}

		/// That record has no parent -> Root 
		/// We are going to use this information to create the nested set representation
		if ($PARENTS==array()){$ROOTS[$ID]=true;}
		else foreach ($PARENTS as $C)
		{
			
			//echo $NAME."\t".$C."\n";
			/// This is causing the creation of a record in $DATA without name/description. See case above
			$DATA[$C]['CHILD'][$ID]=true;
		}

	}
	
fclose($fp);


addLog("Create nested set representation");
	if ($ROOTS==array())														failProcess($JOB_ID."009",'No root found');

/// Create nested set representation that is going to assign boundary numbers.
//// Let's say that the root has for boundary 1 10.
//// The two childs:  A 2-5 and B 6-9
/// And the A has a child C 3-4
/// If we want ALL parents of C, we are going to look outside the boundaries, i.e. <3 for the left side and >4 for the right side.
//// By doing so we get A 2-5 and root 1-10 but not B because the left boundary 6 is above C left boundary.
//// Similarly, if we want children of Root, we will look inside the boundaries i.e >1 for theleft side and <10 for the right side, leading to A B and C.
	$fp=fopen('TREE.csv','w');if (!$fp)													failProcess($JOB_ID."010",'Unable to open TREE.csv');
	function genTree(&$DATA,$ROOTS,$LEVEL,&$LEVEL_V)
	{
		global $fp;
		++$LEVEL;
		foreach ($ROOTS as $RID=>$T)
		{
			if (!isset($DATA[$RID])){echo $RID."\n";continue;}
		
			++$LEVEL_V;$LEVEL_LEFT=$LEVEL_V;
			if (isset($DATA[$RID]['CHILD']))genTree($DATA,$DATA[$RID]['CHILD'],$LEVEL,$LEVEL_V);
			
			++$LEVEL_V;$LEVEL_RIGHT=$LEVEL_V;
			fputs($fp,$DATA[$RID]['DB']."\t".$LEVEL."\t".$LEVEL_LEFT."\t".$LEVEL_RIGHT."\n");
		}
	}
	$VALUE=0;

	genTree($DATA,$ROOTS,0,$VALUE);
	fclose($fp);

	//// $DATA now contains both the data from the database and from the file
	/// So we look at each of those to see what needs to be deleted, updated, inserted
	
	
addLog("Pushing changes to database");

	/// Open the file that is going to contain the new records
	$fpE=fopen('ENTRY.csv','w');if (!$fpE)														failProcess($JOB_ID."011",'Unable to open ENTRY.csv');
	$HAS_NEW_DATA=false;

	/// We are going to look at each record and see what we need to do
	foreach ($DATA as $ID=>$INFO)
	{
		/// FROM_DB => record is not in the file anymore -> DELETE IT	
		if ($INFO['STATUS']=='FROM_DB')
		{
			
			$QUERY=' DELETE FROM bioassay_onto_entry WHERE bioassay_onto_entry_id = '.$INFO['DB'].'';
			if (!runQueryNoRes($QUERY))														failProcess($JOB_ID."012",'Unable to run query '.$QUERY);
		}
		/// Update the record
		else if ($INFO['STATUS']=='TO_UPD')
		{
			
			$QUERY=" UPDATE bioassay_onto_entry SET 
			bioassay_tag_id ='".prepString($ID)."',
			bioassay_label ='".prepString($INFO['NAME'])."',
			bioassay_definition='".prepString(implode("",$INFO['DESC']))."' 
			WHERE bioassay_onto_entry_id = ".$INFO['DB'];
			if (!runQueryNoRes($QUERY))														failProcess($JOB_ID."013",'Unable to run query '.$QUERY);
		}
		/// New record -> in file
		else if ($INFO['STATUS']=='TO_INS')
		{
			$HAS_NEW_DATA=true;
			fputs($fpE,$INFO['DB']."\t".
			$ID."\t".'"'.
			str_replace('"','""',$INFO['NAME']).'"'."\t".'"'.
			str_replace('"','""',implode("",$INFO['DESC'])).'"'."\n");
			//print_r($INFO);
			
		}

		
	}

	fclose($fpE);


addLog("Pushing new records to database");
	if ($HAS_NEW_DATA)
	{
		$command='\COPY '.$GLB_VAR['DB_SCHEMA'].'.bioassay_onto_entry(bioassay_onto_entry_id,bioassay_tag_id,bioassay_label,bioassay_definition)FROM \''."ENTRY.csv"."'  (DELIMITER E'\\t', null \\\"NULL\\\" ,format CSV )";
		echo $DB_INFO['COMMAND'].' -c "'.$command.'"'."\n";
		system($DB_INFO['COMMAND'].' -c "'.$command.'"',$return_code);
		if ($return_code !=0 )																failProcess($JOB_ID."014",'Unable to insert bioassay entry'); 
	}


addLog("delete content of bioassay_onto_hierarchy");

	if (!runQueryNoRes("TRUNCATE TABLE bioassay_onto_hierarchy"))							failProcess($JOB_ID."015",'Unable to truncate bioassay_onto_hierarchy'); 


addLog("Pushing tree to database");
	$command='\COPY '.$GLB_VAR['DB_SCHEMA'].'.bioassay_onto_hierarchy(bioassay_onto_entry_id,bioassay_onto_level,bioassay_onto_level_left,bioassay_onto_level_right)FROM \'TREE.csv'."'  (DELIMITER E'\\t', null \\\"NULL\\\" ,format CSV, HEADER )";
	echo $DB_INFO['COMMAND'].' -c "'.$command.'"'."\n";
	system($DB_INFO['COMMAND'].' -c "'.$command.'"',$return_code);
	if ($return_code !=0 )																	failProcess($JOB_ID."016",'Unable to insert tree'); 



updateStat('bioassay_onto_entry','bioassay_onto',$STATS['ENTRY'],$JOB_ID);
	 


addLog("Delete obsolete files");
 	$list_files=array('TREE.csv','ENTRY.csv');
 	foreach ($list_files as $F)
 	if (is_file($F))unlink($F);

	
addLog("Push to prod");
   pushToProd();
	
	

successProcess();

?>
